deepchecks/vision/base.py

```diff
@@ -362,102 +363,160 @@ def run(
+                checks[check_idx] = copy.deepcopy(check)
             elif isinstance(check, SingleDatasetCheck):
-                if context.train is not None:
-                    new_check: BaseCheck = copy.deepcopy(check)
-                    checks_list.append((str(check_idx) + ' - Train', new_check))
-                if context.test is not None:
-                    new_check: BaseCheck = copy.deepcopy(check)
-                    checks_list.append((str(check_idx) + ' - Test', new_check))
-        check_dict = OrderedDict(checks_list)
-
-        # Initialize all checks
-        for check in check_dict.values():
-            check.initialize_run(context)
-
-        # Create result dict to hold results by check name
-        results = OrderedDict({check_idx: None for check_idx, check in check_dict.items()})
-
-        # Loop over training batches
-        n_batches = len(context.train.get_data_loader())
-        progress_bar = ProgressBar(self.name + ' - Train', n_batches)
-        for batch_id, batch in enumerate(context.train.get_data_loader()):
-            progress_bar.set_text(f'{100 * batch_id / (1. * n_batches):.0f}%')
+                if train_dataset is not None:
+                    checks[str(check_idx) + ' - Train'] = copy.deepcopy(check)
+                if test_dataset is not None:
+                    checks[str(check_idx) + ' - Test'] = copy.deepcopy(check)
+
+        for check in checks.values():
+            if isinstance(check, SingleDatasetCheck):
+                check.initialize_run(
+                    context,
+                    dataset_kind=None # TODO: when this method should be called?
+                )
+            else:
+                check.initialize_run(context)
+
+        results: OrderedDict[
+            Union[str, int],
+            Union[CheckResult, CheckFailure]
+        ] = OrderedDict({})
+
+        run_train_test_checks = train_dataset is not None and test_dataset is not None
+
+        if train_dataset is not None:
+            self._training_loop(
+                checks=checks,
+                data_loader=train_dataset.get_data_loader(),
+                context=context,
+                run_train_test_checks=run_train_test_checks,
+                results=results
+            )
+
+        if test_dataset is not None:
+            self._validation_loop(
+                checks=checks,
+                data_loader=test_dataset.get_data_loader(),
+                context=context,
+                run_train_test_checks=run_train_test_checks,
+                results=results
+            )
+
+        for check_idx, check in checks.items():
+            if check_idx not in results:
+                try:
+                    if isinstance(check, SingleDatasetCheck):
+                        results[check_idx] = check.compute(
+                            context,
+                            dataset_kind=None # TODO: when this method should be called?
```

Comment:
This should be the appropriate dataset

Reply:
I understand this) but this code is outside training/validation loops, which dataset is appropriate?

Reply:
should it be called twice if train and test were passed?

Reply:
fixed, resolving